manager.go

```diff
@@ -0,0 +1,248 @@
+
+package aclstore
+
+import (
+	"context"
+	"log"
+	"net/http"
+	"strings"
+
+	"github.com/julienschmidt/httprouter"
+	"gopkg.in/errgo.v1"
+	httprequest "gopkg.in/httprequest.v1"
+
+	"github.com/juju/aclstore/params"
+)
+
+// Params holds the parameters for a NewManager call.
+type Params struct {
+	// Store holds the persistent storage used by the handler.
+	Store ACLStore
+
+	// RootPath holds the root URL path prefix to use
+	// for the ACL endpoints. All the endpoints will be
+	// prefixed with this path.
+	RootPath string
+
+	// Authenticate authenticates the given HTTP request and returns
+	// the resulting authenticated identity. If authentication
+	// fails, Authenticate is responsible for writing its own
+	// response - the returned error can be used for logging.
+	Authenticate func(ctx context.Context, w http.ResponseWriter, req *http.Request) (Identity, error)
+
+	// InitialAdminUsers holds the contents of the admin ACL
+	// when it is first created.
+	InitialAdminUsers []string
+}
+
+// Identity represents an authenticated user.
+type Identity interface {
+	// Allow reports whether the user should be allowed to access
+	// any of the users or groups in the given ACL slice.
+	Allow(ctx context.Context, acl []string) (bool, error)
+}
+
+// AdminACL holds the name of the administrator ACL.
+const AdminACL = "admin"
+
+// CodeACLNotFound holds the error code returned from
+// the HTTP endpoints when an ACL name has not been
+// created.
+const CodeACLNotFound = "ACL not found"
+
+// Manager implements an ACL manager.
+type Manager struct {
+	p      Params
+	router *httprouter.Router
+}
+
+var errAuthenticationFailed = errgo.Newf("authentication failed")
+
+var reqServer = &httprequest.Server{
+	ErrorWriter: func(ctx context.Context, w http.ResponseWriter, err error) {
+		if errgo.Cause(err) == errAuthenticationFailed {
+			// The Authenticate method has already written its response.
+			return
+		}
+		log.Printf("writing error %#v", err)
+		status, body := errorMapper(ctx, err)
+		httprequest.WriteJSON(w, status, body)
+	},
+}
+
+func errorMapper(ctx context.Context, err error) (int, interface{}) {
+	switch errgo.Cause(err) {
+	case ErrACLNotFound:
+		return http.StatusNotFound, &httprequest.RemoteError{
+			Message: err.Error(),
+			Code:    CodeACLNotFound,
+		}
+	case ErrBadUsername:
+		err = httprequest.Errorf(httprequest.CodeBadRequest, "%v", err)
+	}
+	return httprequest.DefaultErrorMapper(ctx, err)
+}
+
+// NewManager returns a new Manager instance that manages a
+// set of ACLs. It ensures there is at least one ACL
+// created, named "admin", which is given p.InitialAdminUsers
+// when it is first created.
+func NewManager(ctx context.Context, p Params) (*Manager, error) {
+	if err := p.Store.CreateACL(ctx, AdminACL, p.InitialAdminUsers...); err != nil {
+		return nil, errgo.Notef(err, "cannot create initial admin ACL")
+	}
+	m := &Manager{
+		p:      p,
+		router: httprouter.New(),
```

Comment:
by default a new httprouter has a slightly surprising configuration are you sure that's what you want?

Reply:
You should at least overwrite the not-found handler.

Reply:
I've left it as a TODO.